src/hooks/usePinnedEvents.ts

```diff
@@ -0,0 +1,212 @@
+    useEffect(() => {
+        setPinnedEvents(getPinnedEventIds(room));
+        return () => {
+            setPinnedEvents([]);
+        };
+    }, [room]);
+    return pinnedEvents;
+};
+
+/**
+ * Get the read pinned event IDs from a room.
+ * @param room
+ */
+function getReadPinnedEventIds(room?: Room): Set<string> {
+    return new Set(room?.getAccountData(ReadPinsEventId)?.getContent()?.event_ids ?? []);
+}
+
+/**
+ * Get the read pinned event IDs from a room.
+ * @param room
+ */
+export const useReadPinnedEvents = (room?: Room): Set<string> => {
+    const [readPinnedEvents, setReadPinnedEvents] = useState<Set<string>>(new Set());
+
+    // Update the read pinned events when the room state changes
+    // Filter out events that are not read pinned events
+    const update = useCallback(
+        (ev?: MatrixEvent) => {
+            if (ev && ev.getType() !== ReadPinsEventId) return;
+            setReadPinnedEvents(getReadPinnedEventIds(room));
+        },
+        [room],
+    );
+
+    useTypedEventEmitter(room, RoomEvent.AccountData, update);
+    useEffect(() => {
+        setReadPinnedEvents(getReadPinnedEventIds(room));
+        return () => {
+            setReadPinnedEvents(new Set());
+        };
+    }, [room]);
+    return readPinnedEvents;
+};
+
+/**
+ * Fetch the pinned event
+ * @param room
+ * @param pinnedEventId
+ * @param cli
+ */
+async function fetchPinnedEvent(room: Room, pinnedEventId: string, cli: MatrixClient): Promise<MatrixEvent | null> {
+    const timelineSet = room.getUnfilteredTimelineSet();
+    // Get the event from the local timeline
+    const localEvent = timelineSet
+        ?.getTimelineForEvent(pinnedEventId)
+        ?.getEvents()
+        .find((e) => e.getId() === pinnedEventId);
+
+    // Decrypt the event if it's encrypted
+    // Can happen when the tab is refreshed and the pinned events card is opened directly
+    if (localEvent?.isEncrypted()) {
+        await cli.decryptEventIfNeeded(localEvent);
+    }
+
+    // If the event is available locally, return it if it's pinnable
+    // or if it's redacted (to show the redacted event and to be able to unpin it)
+    // Otherwise, return null
+    if (localEvent) return PinningUtils.isUnpinnable(localEvent) ? localEvent : null;
+
+    try {
+        // The event is not available locally, so we fetch the event and latest edit in parallel
+        const [
+            evJson,
+            {
+                events: [edit],
+            },
+        ] = await Promise.all([
+            cli.fetchRoomEvent(room.roomId, pinnedEventId),
+            cli.relations(room.roomId, pinnedEventId, RelationType.Replace, null, { limit: 1 }),
+        ]);
+
+        const event = new MatrixEvent(evJson);
+
+        // Decrypt the event if it's encrypted
+        if (event.isEncrypted()) {
+            await cli.decryptEventIfNeeded(event);
+        }
+
+        // Handle poll events
+        await room.processPollEvents([event]);
```

Comment:
Hmm, really feels like this is pulling logic out of the js-sdk and duplicating it here. Is there any way we could let the js-sdk do all this work?

Reply:
Yeah, I copy pasta this logic from the previous impl of pinning. You are talking about the manual event fetching right?

Reply:
@dbkr My proposition is leave this duplication here since the duplication exists before this PR. And I'll make a dedicated PR to move all the ( https://github.com/matrix-org/matrix-react-sdk/blob/develop/src/components/views/messages/MPollEndBody.tsx#L41 ) duplicate into the js-sdk.

Reply:
Yeah, sounds ideal if you have the time - thank you!